### PR TITLE
Add `Out_channel.{is,set}_buffered` to control buffering of output channels

### PR DESCRIPTION
**Pull request by nojb**
See #6973 

~~The PR currently uses `caml_sys_isatty` to determine whether the channel is connected to a terminal. This is the same function that we use to determine whether to output ANSI color codes, and it returns the right thing on Unix and when running under Cygwin, but currently returns `false` if ran under the Windows console, so that would need fixing if we want to go ahead with this.~~

cc @dra27 @johnwhitington

### CODE REVIEW

**Comment by dra27** [comment]
Given that channels are _part_ of the runtime _exposed_ by the standard library, should this change be in `caml_ml_open_descriptor_out` and thus be inherited by stdlib replacements which also choose to expose OCaml’s runtime support for channels (I’m not sure what I think about this!)

**Comment by johnwhitington** [comment]
My understanding of everything in https://github.com/ocaml/ocaml/issues/6973 is that, if we wanted to match the Unix tradition entirely:

1. stderr should always be unbuffered by default (or just line-buffered? the phrase "not fully buffered" is used) even when not interactive.
2. stdin and stdout should be fully buffered if not interactive, line buffered if interactive.
3. we would want a function to allow the buffered-ness (unbuffered, line buffered, fully buffered) of any `out_channel` to be changed from its default.

However, in order to avoid behaviour changing too much from previous versions of, this PR just adds a function `set_channel_unbuffered`. This approach seems a good one.

But I disagree with this:

>The PR currently uses caml_sys_isatty to determine whether the channel is connected to a terminal. This is the same function that we use to determine whether to output ANSI color codes, and it returns the right thing on Unix and when running under Cygwin.

This means that, if I redirect the output of my (stdout/stderr mixed) program to a file, that file could be different from the terminal output, due to different buffering.

Wouldn't the lowest-surprise change be to simply add the `set_channel_unbuffered` function, and not touch the defaults?

**Comment by nojb** [comment]
> This means that, if I redirect the output of my (stdout/stderr mixed) program to a file, that file could be different from the terminal output, due to different buffering.

Oops sorry, I got mixed up. Indeed `stderr` was meant to be unbuffered regardless of whether it is connected to a terminal or not. (I amended the PR.)

> Wouldn't the lowest-surprise change be to simply add the `set_channel_unbuffered` function, and not touch the defaults?

I see three alternatives:

1. Set `stderr` to be unbuffered unconditionally, no possibility of changing this
2. Set `stderr` to be unbuffered and expose `set_channel_unbuffered` to turn it back to buffered if necessary
3. Expose the function `set_channel_unbuffered` and do not change anything else (your suggestion)

Indeed 3. would be the minimal change...

**Comment by johnwhitington** [comment]
Well, we want both `set_channel_buffered` and `set_channel_unbuffered` for symmetry, surely? I suggest just adding those functions for now, and leaving any change to default settings to the future.

**Comment by nojb** [comment]
> Well, we want both `set_channel_buffered` and `set_channel_unbuffered` for symmetry, surely? I suggest just adding those functions for now, and leaving any change to default settings to the future.

Agree. I added a commit in this direction.

As this touches the stdlib we will need a couple of core developers to formally accept the PR. Maybe @gasche @alainfrisch @dra27 @xavierleroy ?

File: runtime/extern.c
Comment (dra27) [inline runtime/extern.c]: This pattern occurs quite a bit - is it worth having a `CAML_FLUSH_UNBUFFERED(chan);` macro or something?

File: stdlib/stdlib.mli
Comment (dra27) [inline stdlib/stdlib.mli]: At the risk of bike shedding, if we have `set_binary_mode_out`, might this be `set_buffered_mode_out` (or `set_unbuffered_mode_out` - I have no opinion on which way around the bool is)

**Review by dra27 — COMMENTED** [review]
I think this is a good idea - we obviously have to be cautious adding functions to `Stdlib` (alas, still no `IO` module...), but I think this function is a safe name.

I have a small, and potentially irrelevant suggestion for the C code, and a small and potentially bike-sheddy suggestion on the function name.

I think it's a good idea to add the function and then have another PR to make it the default. Do you intend to open that as soon as this is merged for further discussion? Having thought some more, I think it would be a good idea for it to be that `caml_ml_open_open_descriptor_out 2` always returns an unbuffered `out_channel` (i.e. it's done in C, not in OCaml)

File: runtime/io.c
Comment (xavierleroy) [inline runtime/io.c]: You should also flush the channel at this point. Otherwise, if the channel contains buffered data, it will not be visible until the next output operation on the channel, which can be confusing to users.

File: stdlib/stdlib.mli
Comment (xavierleroy) [reply stdlib/stdlib.mli]: Another bit of bike shedding: `set_channel_unbuffered stderr true` sounds like a double negation; what about `set_channel_buffered stderr false` ?  [Edit: actually it's `set_channel_unbuffered stderr false` that is a bit of a double negation.]

**Review by xavierleroy — COMMENTED** [review]
One suggestion and one comment below.

Overall, I think it's a useful addition, and the implementation is reasonably simple.  Ideally, the new function would live in `Stdlib.Out_channel` (see #8937) and not pollute the `Stdlib` name space.

**Comment by nojb** [comment]
> Ideally, the new function would live in `Stdlib.Out_channel` (see #8937) and not pollute the `Stdlib` name space.

I also feel a bit sad about it. I can resucitate #8937 if people are feeling like having a go at it.

File: runtime/io.c
Comment (nojb) [reply runtime/io.c]: Good catch, fixed.

**Comment by nojb** [comment]
I rebased this PR on top of #10545 and moved the function to `Out_channel`.

The signature is currently `set_buffering_mode: out_channel -> bool -> unit` which is the same as `set_binary_mode`, but am not sure how well it reads. Perhaps `set_buffered_mode` would be better?

@xavierleroy would you mind reviewing the PR again? Thanks!

File: runtime/extern.c
Comment (nojb) [reply runtime/extern.c]: Thanks David for the suggestion. I did this (the macro is called `Flush_if_unbuffered`).

**Comment by nojb** [comment]
> I think it's a good idea to add the function and then have another PR to make it the default. Do you intend to open that as soon as this is merged for further discussion?

Personally I don't have a strong opinion as to whether `stderr` should be unbuffered by default. If people think that should be the case, please chime in; I can submit the (trivial) PR to do that.

**Comment by johnwhitington** [comment]
In #6973, @pmetzger wrote:

> stderr is definitely conventionally unbuffered. The goal is that error output is rare, and should show up immediately, both so the programmer knows when a problem happened, and so that a crash will not deprive the programmer of debugging information.

which I find pretty convincing (both the convention part and the deprivation of output part).

**Comment by nojb** [comment]
Friendly ping. We need two formal reviews for this one. I am counting on @xavierleroy to do one, since he has already done it for a previous version. Perhaps @dra27 can do the second one? It is a small amount of code, probably the hardest part is to agree on the name...

One point I was wondering about is that currently there is no way to find out if the channel is buffered or not. We could make the new function return the "old" value, ie have a signature `out_channel -> bool -> bool` instead of `out_channel -> bool -> unit`. This could make it possible to modify the buffering mode of a channel in a block, restoring its old setting at the end (would this be useful in practice?).

**Comment by dbuenzli** [comment]
`Out_channel.set_buffering_mode true` does't read that great in my opinion. 

What about: 
```ocaml
Out_channel.buffered : out_channel -> bool
Out_channel.set_buffered : out_channel -> bool -> unit
```

**Comment by nojb** [comment]
> `Out_channel.set_buffering_mode true` does't read that great in my opinion.
> 
> What about:
> 
> ```ocaml
> Out_channel.buffered : out_channel -> bool
> Out_channel.set_buffered : out_channel -> bool -> unit
> ```

Sounds reasonable. To avoid too much back-and-forth, let's wait a bit before going ahead with this choice in case there are other opinions.

File: runtime/io.c
Comment (dra27) [inline runtime/io.c]: IIUC, it's OK to update the channel's flags without registering a root on `vchannel` and locking the channel, but I think you need to do have done both in order to call `caml_flush` on it. Simplest to call `caml_ml_flush` therefore?

**Comment by dra27** [comment]
I agree with @dbuenzli's `set_buffered` and I guess there's a reason for `buffered` rather than `is_buffered`?

`Stdlib.set_binary_mode_out` has more need of the `_mode_` than `Out_channel.set_binary` (if that wanted to be changed prior to 4.14) and therefore similarly for `set_buffered`.

File: runtime/io.c
Comment (nojb) [reply runtime/io.c]: Good catch. Fixed.

**Comment by nojb** [comment]
> I agree with @dbuenzli's `set_buffered` and I guess there's a reason for `buffered` rather than `is_buffered`?

OK I added `is_buffered` (I kept the `is_` because that seems to be the style used in most other modules, eg `Uchar.is_valid`).

> `Stdlib.set_binary_mode_out` has more need of the `_mode_` than `Out_channel.set_binary` (if that wanted to be changed prior to 4.14) and therefore similarly for `set_buffered`.

Yes, I was wondering about that... we should make up our minds before the 4.14 release...

**Comment by dbuenzli** [comment]
> I agree with @dbuenzli's `set_buffered` and I guess there's a reason for `buffered` rather than `is_buffered`?

Not really `is_buffered` is fine with me (for a mutable property `p` I tend to have `M.p` and `M.set_p` but don't mind making exceptions for booleans).

**Review by damiendoligez — APPROVED** [review]
Have you given any thought to the question of whether to add some tests ?

**Comment by pmetzger** [comment]
Testing buffered vs. unbuffered behavior in a principled way is a little difficult to do without playing with ptys.

File: runtime/io.c
Comment (damiendoligez) [inline runtime/io.c]:
```suggestion
  return Val_bool( ! (channel->flags & CHANNEL_FLAG_UNBUFFERED));
```

**Review by damiendoligez — APPROVED** [review]
Sorry, I meant to have this nitpick in my review but somehow failed to save it.

**Comment by damiendoligez** [comment]
> Testing buffered vs. unbuffered behavior in a principled way is a little difficult to do without playing with ptys.

Do some write, then commit suicide with a self-sigKILL ? Yuck.
I won't insist on adding tests.

File: runtime/io.c
Comment (nojb) [reply runtime/io.c]: Thanks, fixed.

**Comment by nojb** [comment]
> Have you given any thought to the question of whether to add some tests ?

I pushed a test that works by comparing the behaviour between stdout and stderr, where I set one of them to be unbuffered. It is not ideal, but it tests some of the functionality and is portable :) Thanks!

**Comment by pmetzger** [comment]
> > Testing buffered vs. unbuffered behavior in a principled way is a little difficult to do without playing with ptys.
> 
> Do some write, then commit suicide with a self-sigKILL ? Yuck.
> I won't insist on adding tests.

That's probably for the best, yes. That said, OCaml library tools for dealing with ptys on architectures that support them might be an interesting thing to do someday (someday ≠ any time soon.)

**Comment by nojb** [comment]
> > I won't insist on adding tests.
> 
> That's probably for the best, yes. 

Note that I *did* add some (granted, not ideal) tests

**Review by xavierleroy — APPROVED** [review]
I read the diff again and it looks very good to me.  Thanks!

**Pull request merged**
